app/main/views/suppliers.py

```diff
@@ -86,6 +89,116 @@ def download_agreement_file(supplier_id, framework_slug, document_name):
+        supplier=supplier,
+        framework=framework,
+        countersigned_agreement=countersigned_agreement,
+        **get_template_data())
+
+
+@main.route('/suppliers/<supplier_id>/countersigned-agreements/<framework_slug>/'
+            '<document_name>', methods=['POST'])
+@login_required
+@role_required('admin-ccs-sourcing')
+def upload_countersigned_agreement_file(supplier_id, framework_slug, document_name):
+    agreements_bucket = s3.S3(current_app.config['DM_AGREEMENTS_BUCKET'])
+    errors = {}
+
+    if request.files.get('countersigned_agreement'):
+        the_file = request.files['countersigned_agreement']
+        if not file_is_pdf(the_file):
+            errors['countersigned_agreement'] = 'not_pdf'
+
+        if 'countersigned_agreement' not in errors.keys():
+            filename = '{}/agreements/{}/{}-countersigned-framework-agreement.pdf'.format(
+                framework_slug, supplier_id, supplier_id)
+            agreements_bucket.save(filename, the_file)
+
+            data_api_client.create_audit_event(
+                audit_type=AuditTypes.upload_countersigned_agreement,
+                user=current_user.email_address,
+                object_type='suppliers',
+                object_id=supplier_id,
+                data={'upload_countersigned_agreement': filename})
+
+            flash('countersigned_agreement', 'upload_countersigned_agreement')
+
+    if len(errors) > 0:
+        for category, message in errors.items():
+            flash(category, message)
+
+    return redirect(url_for(
+        '.list_countersigned_agreement_file',
+        supplier_id=supplier_id,
+        framework_slug=framework_slug,
+        document_name=document_name)
+    )
+
+
+@main.route('/suppliers/<supplier_id>/countersigned-agreements_download/<framework_slug>/'
+            '<document_name>', methods=['GET'])
+@login_required
+@role_required('admin-ccs-sourcing')
+def download_countersigned_agreement_file(supplier_id, framework_slug, document_name):
+    agreements_bucket = s3.S3(current_app.config['DM_AGREEMENTS_BUCKET'])
+    path = get_countersigned_agreement_document_path(framework_slug, supplier_id)
+    countersigned_agreement_document = agreements_bucket.get_key(path)
+    if not len(countersigned_agreement_document):
+        abort(404)
+    url = get_signed_url(agreements_bucket, countersigned_agreement_document, current_app.config['DM_ASSETS_URL'])
+    if not url:
+        abort(404)
+    return redirect(url)
+
+
+@main.route('/suppliers/<supplier_id>/countersigned-agreements_remove/<framework_slug>/'
+            '<document_name>', methods=['GET', 'POST'])
+@login_required
+@role_required('admin-ccs-sourcing')
+def remove_countersigned_agreement_file(supplier_id, framework_slug, document_name):
+    agreements_bucket = s3.S3(current_app.config['DM_AGREEMENTS_BUCKET'])
+    document = get_countersigned_agreement_document_path(framework_slug, supplier_id)
+
+    if request.method == 'GET':
```

Comment:
This seems like a really good way of doing the 'are you sure' confirmation thing.